src/DynamoCoreWpf/Rendering/HelixRenderPackage.cs

```diff
@@ -775,34 +772,174 @@ public List<int> TextureMapsStrideList
+        /// <param name="m42"></param>
+        /// <param name="m43"></param>
+        /// <param name="m44"></param>
+        public void AddInstanceMatrix(float m11, float m12, float m13, float m14,
+            float m21, float m22, float m23, float m24,
+            float m31, float m32, float m33, float m34,
+            float m41, float m42, float m43, float m44, Guid id)
+        {
+            if (!ContainsTessellationId(id))
+            {
+                throw new Exception("The reference to the graphics range(mesh or line) for this ID does not exists.");
+            }
+
+            var transform = new Matrix(m11, m12, m13, m14, m21, m22, m23, m24, m31, m32, m33, m34, m41, m42, m43, m44);
+
+            List<Matrix> transforms;
+            if (instanceTransforms.TryGetValue(id, out transforms))
+            {
+                transforms.Add(transform);
+            }
+            else
+            {
+                instanceTransforms.Add(id, new List<Matrix> {transform});
+            }
+        }
+
+        /// <summary>
+        /// Set the transform as a double array, this transform is applied to all geometry in the renderPackage.
+        /// This matrix should be laid out as follows in row vector order:
+        /// [Xx,Xy,Xz, 0,
+        ///  Yx, Yy, Yz, 0,
+        ///  Zx, Zy, Zz, 0,
+        ///  offsetX, offsetY, offsetZ, W]
+        /// NOTE: The caller of this method should transform the matrix from row vector order to whatever form is needed by the implementation.
+        /// When converting from ProtoGeometry CoordinateSystem form to input matrix, set the first row to the X axis of the CS,
+        /// the second row to the Y axis of the CS, the third row to the Z axis of the CS, and the last row to the CS origin, where W = 1. 
+        /// </summary>
+        /// <param name="matrix"></param>
+        public void AddInstanceMatrix(float[] matrix, Guid id) 
+        {
+            if (!ContainsTessellationId(id))
+            {
+                throw new Exception("The reference to the graphics range(mesh or line) for this ID does not exists.");
+            }
+
+            var transform = new Matrix(matrix);
+
+            List<Matrix> transforms;
+            if (instanceTransforms.TryGetValue(id, out transforms))
+            {
+                transforms.Add(transform);
+            }
+            else
+            {
+                instanceTransforms.Add(id, new List<Matrix> { transform });
+            }
+        }
+
+        /// <summary>
+        /// Checks if a base tessellation guid has already been registered with this <see cref="IInstancingRenderPackage"/>.
+        /// Both Line and Mesh ids are checked.
+        /// </summary>
+        /// <param name="id"></param>
+        /// <returns></returns>
+        public bool ContainsTessellationId(Guid id)
+        {
+            return MeshVertexRangesAssociatedWithInstancing.ContainsKey(id) || LineVertexRangesAssociatedWithInstancing.ContainsKey(id);
```

Comment:
@mjkkirschner Hmmmmm.... Not sure but we may need to handle a transform that has both mesh and line type tessellation data.  Maybe when edges are on.  I can't remember how that case was supported.

Reply:
It might be a bit of a hack but the LIbG portion could actually handle this ^^^ I think.  I always wondered why we didn't handle the `ShowEdges` in the base tessellation call.  Maybe puts too much ownership on those implementing IGraphicItem.  In this case we are still internal so it may work to let the instanceableItem handle the data.  In the case of a cube, `AddBaseTessellation(instancingPackage, factory.TessellationParameters)` could inspect the `TessellationParameters` and add only a mesh or a mesh and line data edges are needed.  They would have to have separate GUIDs.  The subsequent call to `AddInstance(instancingPackage, factory.TessellationParameters, labelKey)` could also add the correct instance references fo either mesh only or mesh and line data.

Reply:
This may also be follow up task?

Reply:
⚠️ I'll take a quick look but ,if nothing obvious and simple I think I'll have to file a followup as the allotted time for this task is running out.

I wonder if we can just push the line data to the baseTessellation from this method:
https://github.com/DynamoDS/Dynamo/blob/9740b24b50257b890bafd9d2efdb9904690c32b2/src/DynamoCore/Scheduler/UpdateRenderPackageAsyncTask.cs#L262

or always generate the edge data in baseTessellation but only using it (extracting it from the render package) if show edges is on.

For the current state of the PR, I think we need to decide the following:
1. When show_edges is on, we don't use instancing OR -
2. When show-edges is on, we use instancing but don't draw edges...

Reply:
Could be also...
4. Edges added within instancing implementation.